File: mobile/lib/core/services/socket_service.dart
import 'package:google_maps_flutter/google_maps_flutter.dart';
import 'package:socket_io_client/socket_io_client.dart' as IO;
import '../config/app_config.dart';

/// Socket Service for real-time shipper location tracking
/// Inspired by Flutter-Delivery-App reference project
class SocketService {
  static SocketService? _instance;
  IO.Socket? _socket;
  bool _isConnected = false;

  SocketService._();

  static SocketService get instance {
    _instance ??= SocketService._();
    return _instance!;
  }

  bool get isConnected => _isConnected;

  /// Initialize socket connection
  void connect() {
    if (_socket != null && _isConnected) return;

    _socket = IO.io(
      AppConfig.wsBaseUrl,
      IO.OptionBuilder()
          .setTransports(['websocket'])
          .enableAutoConnect()
          .enableReconnection()
          .setReconnectionAttempts(5)
          .setReconnectionDelay(1000)
          .build(),
    );

    _socket!.onConnect((_) {
      _isConnected = true;
      print('SocketService: Connected');
    });

    _socket!.onDisconnect((_) {
      _isConnected = false;
      print('SocketService: Disconnected');
    });

    _socket!.onConnectError((error) {
      print('SocketService: Connection error - $error');
    });

    _socket!.connect();
  }

  /// Disconnect socket
  void disconnect() {
    _socket?.disconnect();
    _socket?.dispose();
    _socket = null;
    _isConnected = false;
  }

  /// Emit shipper location to server (for shipper app)
  /// Server will broadcast to clients tracking this shipment
  void emitShipperLocation({
    required String shipmentId,
    required String shipperId,
    required double latitude,
    required double longitude,
    double? heading,
    double? speed,
  }) {
    if (_socket == null || !_isConnected) {
      print('SocketService: Not connected, cannot emit location');
      return;
    }

    _socket!.emit('shipper:location', {
      'shipmentId': shipmentId,
      'shipperId': shipperId,
      'latitude': latitude,
      'longitude': longitude,
      'heading': heading ?? 0,
      'speed': speed ?? 0,
      'timestamp': DateTime.now().toIso8601String(),
    });
  }

  /// Join a shipment room to receive location updates (for customer tracking)
  void joinShipmentTracking(String shipmentId) {
    if (_socket == null || !_isConnected) return;
    _socket!.emit('tracking:join', {'shipmentId': shipmentId});
  }

  /// Leave shipment tracking room
  void leaveShipmentTracking(String shipmentId) {
    if (_socket == null || !_isConnected) return;
    _socket!.emit('tracking:leave', {'shipmentId': shipmentId});
  }

  /// Listen for shipper location updates (for customer tracking)
  void onShipperLocationUpdate(
    String shipmentId,
    void Function(ShipperLocationUpdate) callback,
  ) {
    if (_socket == null) return;

    _socket!.on('shipper:location:$shipmentId', (data) {
      try {
        final update = ShipperLocationUpdate.fromJson(data);
        callback(update);
      } catch (e) {
        print('SocketService: Error parsing location update - $e');
      }
    });
  }

  /// Remove listener for shipper location updates
  void offShipperLocationUpdate(String shipmentId) {
    _socket?.off('shipper:location:$shipmentId');
  }

  /// Listen for shipment status updates
  void onShipmentStatusUpdate(void Function(ShipmentStatusUpdate) callback) {
    if (_socket == null) return;

    _socket!.on('shipment:status', (data) {
      try {
        final update = ShipmentStatusUpdate.fromJson(data);
        callback(update);
      } catch (e) {
        print('SocketService: Error parsing status update - $e');
      }
    });
  }

  /// Authenticate shipper after login
  void authenticateShipper(String shipperId, String token) {
    if (_socket == null || !_isConnected) return;
    _socket!.emit('shipper:auth', {
      'shipperId': shipperId,
      'token': token,
    });
  }
}

/// Shipper location update model
class ShipperLocationUpdate {
  final String shipmentId;
  final String shipperId;
  final LatLng location;
  final double heading;
  final double speed;
  final DateTime timestamp;

  ShipperLocationUpdate({
    required this.shipmentId,
    required this.shipperId,
    required this.location,
    required this.heading,
    required this.speed,
    required this.timestamp,
  });

  factory ShipperLocationUpdate.fromJson(Map<String, dynamic> json) {
    return ShipperLocationUpdate(
      shipmentId: json['shipmentId'] ?? '',
      shipperId: json['shipperId'] ?? '',
      location: LatLng(
        (json['latitude'] as num?)?.toDouble() ?? 0,
        (json['longitude'] as num?)?.toDouble() ?? 0,
      ),
      heading: (json['heading'] as num?)?.toDouble() ?? 0,
      speed: (json['speed'] as num?)?.toDouble() ?? 0,
      timestamp: json['timestamp'] != null
          ? DateTime.parse(json['timestamp'])
          : DateTime.now(),
    );
  }
}

/// Shipment status update model
class ShipmentStatusUpdate {
  final String shipmentId;
  final String status;
  final String? message;
  final DateTime timestamp;

  ShipmentStatusUpdate({
    required this.shipmentId,
    required this.status,
    this.message,
    required this.timestamp,
  });

  factory ShipmentStatusUpdate.fromJson(Map<String, dynamic> json) {
    return ShipmentStatusUpdate(
      shipmentId: json['shipmentId'] ?? '',
      status: json['status'] ?? '',
      message: json['message'],
      timestamp: json['timestamp'] != null
          ? DateTime.parse(json['timestamp'])
          : DateTime.now(),
    );
  }
}

File: backend/src/shared/socket/socket.service.js
/**
 * Socket.io Service for Real-time Communication
 * Handles shipper location tracking and shipment status updates
 * 
 * Inspired by Flutter-Delivery-App reference project
 */

const { Server } = require('socket.io');
const jwt = require('jsonwebtoken');

let io = null;

/**
 * Initialize Socket.io server
 * @param {http.Server} httpServer - HTTP server instance
 */
function initializeSocket(httpServer) {
  io = new Server(httpServer, {
    cors: {
      origin: '*', // In production, restrict to specific origins
      methods: ['GET', 'POST'],
    },
    transports: ['websocket', 'polling'],
  });

  // Authentication middleware
  io.use(async (socket, next) => {
    try {
      const token = socket.handshake.auth.token || socket.handshake.query.token;
      
      if (token) {
        const decoded = jwt.verify(token, process.env.JWT_SECRET);
        socket.userId = decoded.userId;
        socket.userRole = decoded.role;
      }
      
      next();
    } catch (error) {
      // Allow connection without auth for tracking (customers don't need to login)
      next();
    }
  });

  io.on('connection', (socket) => {
    console.log(`Socket connected: ${socket.id}`);

    // Shipper authenticates after login
    socket.on('shipper:auth', (data) => {
      const { shipperId, token } = data;
      try {
        const decoded = jwt.verify(token, process.env.JWT_SECRET);
        socket.shipperId = shipperId || decoded.userId;
        socket.join(`shipper:${socket.shipperId}`);
        console.log(`Shipper authenticated: ${socket.shipperId}`);
      } catch (error) {
        console.error('Shipper auth failed:', error.message);
      }
    });

    // Shipper emits location update
    socket.on('shipper:location', (data) => {
      const { shipmentId, shipperId, latitude, longitude, heading, speed, timestamp } = data;
      
      // Broadcast to all clients tracking this shipment
      io.to(`tracking:${shipmentId}`).emit(`shipper:location:${shipmentId}`, {
        shipmentId,
        shipperId: shipperId || socket.shipperId,
        latitude,
        longitude,
        heading,
        speed,
        timestamp: timestamp || new Date().toISOString(),
      });
    });

    // Customer joins tracking room for a shipment
    socket.on('tracking:join', (data) => {
      const { shipmentId } = data;
      socket.join(`tracking:${shipmentId}`);
      console.log(`Client ${socket.id} joined tracking:${shipmentId}`);
    });

    // Customer leaves tracking room
    socket.on('tracking:leave', (data) => {
      const { shipmentId } = data;
      socket.leave(`tracking:${shipmentId}`);
      console.log(`Client ${socket.id} left tracking:${shipmentId}`);
    });

    socket.on('disconnect', () => {
      console.log(`Socket disconnected: ${socket.id}`);
    });
  });

  console.log('Socket.io initialized');
  return io;
}

/**
 * Get Socket.io instance
 */
function getIO() {
  if (!io) {
    throw new Error('Socket.io not initialized. Call initializeSocket first.');
  }
  return io;
}

/**
 * Emit shipment status update to all tracking clients
 * @param {string} shipmentId - Shipment ID
 * @param {string} status - New status
 * @param {string} message - Optional message
 */
function emitShipmentStatusUpdate(shipmentId, status, message = null) {
  if (!io) return;
  
  io.to(`tracking:${shipmentId}`).emit('shipment:status', {
    shipmentId,
    status,
    message,
    timestamp: new Date().toISOString(),
  });
}

/**
 * Emit shipper location update (called from backend when location is saved)
 * @param {string} shipmentId - Shipment ID
 * @param {object} location - Location data
 */
function emitShipperLocation(shipmentId, location) {
  if (!io) return;
  
  io.to(`tracking:${shipmentId}`).emit(`shipper:location:${shipmentId}`, {
    shipmentId,
    ...location,
    timestamp: new Date().toISOString(),
  });
}

module.exports = {
  initializeSocket,
  getIO,
  emitShipmentStatusUpdate,
  emitShipperLocation,
};
